Improve RPM and DEB packages for multi-user Nix

abbradar opened the pull request:
This is an effort to make installation of Nix on Debian and Red Hat-based distributions much easier experience. As an end result, you can now install .deb or .rpm built by Hydra and have store created, a minimal nix.conf, build users for Nix daemon created and other changes so you can just enable/start nix-daemon, relogin (to apply a profile script) and use multi-user Nix. Also this fixes package dependencies and makes other improvements.

One of the bigger changes is move from checkinstall to full-featured Debian packaging. Checkinstall has several limitations that are not nice to work with, particularly its configuration files detection and inability to detect `/nix/store` directory creation. Debian packaging is much more complex than I thought, but this gives us ability to fully control all aspects of package.

Also there is a general change that affects overall Nix -- I've replaced underscore in Nix unstable version with dot. This is because Debian explicitly forbids usage of underscores in versions. I'm not sure if this is a good idea -- thoughts?

This was tested in both Docker containers (it was my first encounter with it and it appears to be very useful for such development) and full-fledged VMs (for daemon testing) for Fedora 24 and Ubuntu 16.04. Test plan:

1. Install a package;
2. Enable/start daemon (distro-specific, for systemd: `systemctl enable nix-daemon; systemctl start nix-daemon`);
3. Re-login;
4. Multi-user Nix should work.

abbradar commented:
You also need https://github.com/NixOS/nixpkgs/pull/20767 for this to work.

matthewbauer commented on release.nix:
```diff
@@ -296,26 +298,40 @@ let
+        make installcheck
+      '';
+
+      installPhase = ''
```
Just a nitpick, but this seems like this could be split into patch, install, and build phases. Can we maybe even move this logic back into "debBuild"?

matthewbauer commented on release.nix:
```diff
@@ -296,26 +298,40 @@ let
+
+      installPhase = ''
+        # Hack to set desired init system.
+        sed -i 's,--with-init-system=systemd,--with-init-system=${initSystem},g' debian/rules
```
You could alternatively have a make rule in debian/rules like:

```make
INIT_SYSTEM ?= systemd
```

Then we can just run "export INIT_SYSTEM = initSystem" here.

abbradar replied:
Much better, thanks.

matthewbauer commented on nix.spec.in:
```diff
@@ -118,52 +124,27 @@ make DESTDIR=$RPM_BUILD_ROOT install
-systemctl enable nix-daemon.socket nix-daemon.service
-systemctl start  nix-daemon.socket
-%endif
+%(cat debian/postinst)
```
Can we make "postinst" more generic? Maybe put it into ./scripts/ as "install-groups-linux.sh.in" so we can use it elsewhere?

matthewbauer commented on debian/changelog.in:
```diff
@@ -0,0 +1,5 @@
+
+  * Initial release.
+
+ -- Eelco Dolstra <eelco.dolstra@logicblox.com>  Sun, 13 Dec 2009 00:11:46 +0300
```
Should this date be the date of the initial release or the data of PACKAGE_VERSION release?

matthewbauer commented:
Another question: do users of Nix still need to have ```. $HOME/.nix-profile/etc/profile.d/nix.sh``` in their profile or is there a way to get it setup for everyone?

abbradar commented:
Hm, I think this is not needed now -- `/etc/profile.d/nix.sh` does everything that's needed.

abbradar commented on scripts/nix.conf:
```diff
@@ -0,0 +1,4 @@
+# Default Nix configuration file.
+binary-caches = https://cache.nixos.org/
+signed-binary-caches = *
+binary-cache-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
```
It's the default, so no need to set it explicitly.

abbradar commented:
I'm not sure where to take date from. We can either set it to some pre-defined meaningful date, or somehow derive it from git commit information or wall clock (I don't like either of those because of nondeterminism/reliance on git)...

abbradar commented:
I've settled for timestamp of commit b07ba22 as the datetime in changelog. Also resolved merge conflicts. This is ready for another round of review/merge.

abbradar commented:
I've pushed new version which also fixes Debian packages while making things simpler and less hacky. Notice that this introduces new command to nix-store -- `--print-store-dir` (which, as the title says, prints the store dir -- usually `/nix/store`).

abbradar commented:
I've resolved conflicts again but I don't want to keep resolving them once a month while awaiting a review. I hope it's okay to just leave it as is -- if it comes out of sync again please ping me and I'll resolve it!

abbradar commented:
Ping?

therealpxc commented:
How many more ought to examine this before it will be considered mergeable?

I can't offer that since I'm not a core contributor or whatever, but I'd like to try this thing out. It feels painful to even ask, with how long this has been sitting here, but would you mind resolving the current merge conflicts so that I can play with this, @abbradar ?

abbradar commented:
@therealpxc You are welcome; I've rebased this (added some new fixes, too). I haven't mostly tested this though (.deb builds fine but .rpm fails with a virtual machine memory error, which I assume is my host machine problem).

therealpxc commented:
Thanks! 

I'll see if I can get the RPM building on my machine at home today and let you know if I run into any issues. :-)

therealpxc commented:
So the RPM does indeed  build. I've been unable to test it, however, because I don't have a Fedora 25 machine around. On openSUSE 42.2, the RPM is uninstallable because of the dependency on `libcrypto.so.10`. The machine I really need this for running RHEL 6 anyway, though (I don't know what point release).

Do you happen to know if a similar strategy to the one we're using to build Nix for Fedora 25 can be used for RHEL?

abbradar commented:
@therealpxc Actually we are using Nix in production on CentOS 7 servers (which would correspond to RHEL 7). For this you'd need to use stable Nix which doesn't use C++14. Hopefully you can also build it for RHEL6. I have a branch with my patches rebased atop Nix 1.11 for that; I can find it and publish for you later.

therealpxc commented:
That would be wonderful. Thanks for the clarification on the state of things.

therealpxc commented:
@abbradar I tried building based on this branch for RHEL6 (6.1 and 6.5) and it complained about missing `libseccomp-devel`. Is this due to the version of Nix being too new?

abbradar commented:
@therealpxc Sorry, I have forgotten about this and then had a vacation.

I've rebased my 1.11 branch to include new changes and tried to build it for CentOS 6.5. Unfortunately it doesn't have C++11-compliant compiler... It seems you'd need to build it by yourself on your target system using a backported GCC (I'm sure there are some prebuilt in the wild, not so sure about their authenticity).

FWIW here is my branch: https://github.com/abbradar/nix/tree/stable-1.11

therealpxc commented:
Thanks for coming back to this! I was wondering when and how it would be polite to poke you about this. :-)

I'm already using Nix on that CentOS system in a `pivot_root` environment, so I can probably even use that Nix installation, whose main defect is slowness, to get sufficiently recent GCC working and build it that way.

I'm excited to give this a try ASAP. I'd also like to try to get RPMs working for recent openSUSE and do some general testing for you on this PR to help nudge it forward.

Thanks again!

abbradar commented:
@therealpxc A small advice -- you'd need to build it statically linked, otherwise you'd have references to GCC in the Nix store in the result.

therealpxc commented on release.nix:
```diff
@@ -17,7 +17,7 @@ let
         name = "nix-tarball";
         version = builtins.readFile ./version;
-        versionSuffix = if officialRelease then "" else "pre${toString nix.revCount}_${nix.shortRev}";
+        versionSuffix = if officialRelease then "" else "pre${toString nix.revCount}.${nix.shortRev}";
```
The convention I've seen used for this in other Debian and Ubuntu packages is `<normal_version_number>+git<little endian date>.<time>.<hash-prefix>`, e.g., [`0.2.0+git20170602.1830.1cb3fb`](https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers/+packages).

abbradar replied:
We sadly don't have date and time information here and I don't want to change the naming schema much (we are distribution-agnostic after all, or is this more common than Debian derivatives?).

edolstra replied:
Why is the version format changed?

abbradar replied:
This is because Debian explicitly forbids usage of underscores in versions:
```
dpkg-buildpackage: warning:     debian/changelog(l1): version '2.0pre0_0000000' is invalid: version number contains illegal character '_'
```

therealpxc commented:
Hey, @abbradar.

Because I've been lazy/busy, I didn't get into trying to manually build the RPMs last night, and I'm happy to blaze that trail alone, but I have some questions about the other stuff and some other concerns.

One is that the ole `curl|bash` installer doesn't work on CentOS 6, so building the RPMs myself seems yet more attractive, if difficult for the reasons you anticipated. I may file a bug for that soon, but I might be on my own in figuring it out.

The other is that the CentOS 7 RPM doesn't build for me, it gives a compile error. Could you take a look at the CentOS 7 RPM of the stable branch and at least tell me if you have the same issue?

gilligan commented:
@abbradar what is the status on this PR?

abbradar commented:
@gilligan I've took some time to redo it after NixCon where we discussed it with @edolstra. It's finished now!

@edolstra ping! This PR is now ready; packages work at least on Fedora 25 and Ubuntu 16.04. We also have tests now for those packages.

abbradar commented:
@therealpxc Sorry, I have missed your old message ;_; I haven't tried building RPMs for Cent OS 7 now, I'll see if they work in the following week. It builds on Fedora 25 FWIW.

bjornfor commented on scripts/nix-profile.sh.in:
```diff
@@ -1,6 +1,6 @@
 if [ -n "$HOME" ] && [ -n "$USER" ]; then
     __savedpath="$PATH"
-    export PATH=@coreutils@
+    export PATH=@coreutils@:/usr/bin:/bin
```
I was (is) confused about this too. Can you say something in the commit message about why /bin etc. must be added to PATH?

abbradar replied:
@bjornfor Done.

edolstra replied:
The better way would be to have configure look for the path of the other coreutils (any of them will do).

abbradar replied:
I'm a bit stuck how to fix this completely -- in various places in code we use `@coreutils@` to get their paths but on Debian they are semi-randomly distributed between `/bin` and `/usr/bin`. For example, what should `coreutils` in `config.nix` be equal to?

abbradar commented:
@therealpxc I got myself some extra time so I've tried building it on CentOS 7. Unfortunately because CentOS 7.3 doesn't have a C++14 capable compiler we can't build a package using stock compiler.

bhipple commented:
This is fantastic, thanks abbradar!  There was a lot of talk at NixCon about encouraging wider enterprise adoption for NixPkgs, and providing RPMs / DEBs that a sysadmin can install to get multi-user nix on their servers via familiar configuration management tools is a huge step in that direction.

edolstra commented on scripts/nix.conf:
```diff
@@ -0,0 +1,4 @@
+# Default Nix configuration file.
+binary-caches = https://cache.nixos.org/
+signed-binary-caches = *
+binary-cache-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
```
It's fine to install an empty configuration file, but it must not duplicate Nix's built-in defaults. So all the settings here should be removed.

Also, `binary-caches` has been renamed to `substituters`.

abbradar replied:
Ah, fair enough -- let's remove it altogether then.

abbradar replied:
On the second thought I've left it with an advice to read nix.conf manpage.

bhipple replied:
While it may seem insane to do so from the perspective of an individual user, it's not uncommon for many enterprise environments to prefer mirroring and building everything from source in-house for security / auditing / reproducibility-without-external-internet, among other reasons.  Of course, it's certainly feasible for them to maintain a patched fork of nix, but nevertheless it'd be nice if the cache options were entirely in the configuration file instead of hard-coded in C++.  There also [a few users](https://github.com/NixOS/nixpkgs/issues/20542) who get confused when a `nix.conf` with 0 keys and 0 binary caches still substitutes from cache.nixos.org.

Just throwing this out there as food for thought; I'd imagine it's outside the scope of this particular PR.

abbradar replied:
This is out of scope yeah, we can include `nix.conf.example` with commented options explaining what each does and what the defaults are.

edolstra replied:
This file still needs to be removed.

abbradar replied:
Fixed.

edolstra commented on scripts/post-install-linux.sh:
```diff
@@ -0,0 +1,31 @@
+  groupadd -r "$nixbld_group"
+fi
+
+for i in $(seq 10); do
```
The default of 10 is insufficient for modern systems. NixOS creates 32 build users.

williammpratt replied:
Why not just compute the number? A mathematical upper bound is the multiplication of all constants in nix.conf. What the lowest upper bound is, is an exercise left for the reader. 

Hard coding a constant to work for all systems won't work in general.

abbradar replied:
Missed this one. I've bumped the limit to 32.

@williammpratt We can't really compute it anyhow -- this number specifies (indirectly) maximum number of concurrent builds, because each build is run from its own user. Therefore the only requirement for this number is to be bigger or equal than `max-jobs` and even that may be not enough. 32 is big enough for most cases so we use it.

edolstra commented on src/nlohmann/json.hpp:
```diff
@@ -0,0 +1,13003 @@
+/*
```
Why does this file need to be added to the repo? It's already added to the tarball in `local.mk`.

abbradar replied:
The problem is that currently you can't build Git checkout of Nix on a system without Nix, and the only thing which breaks that is json.hpp because it's added from nixpkgs. With this I can build Nix from scratch in Ubuntu/Fedora Docker container.

FRidh replied:
This change has landed in master 7f2c324ed18cba4004ff89dfd84cf2df979b2571

therealpxc commented:
Is there a reason not to enhance the Nix expressions we use for building RPMs so that additional repositories (like some containing a newer GCC) can be used instead of just relying on a single repomd.xml.gz file? I think that could help us stay ahead of problems like this in the future, since RedHat usually have special repos for this purpose, like the devtoolset repos.

I am interested in doing this and intend to look into implementing it unless someone better informed can tell me why that approach is ill-advised. What do you think?

bhipple commented:
@therealpxc that sounds like a fantastic idea; I was considering doing something like that myself.  I'm trying to find what the cleanest way to accomplish that is.  Perhaps we should add a way to install and enable Software Collections via the `scl` utility to get access to C++14 and libraries like `libsodium`?  Both of these are necessary to do a master branch build of `nix`.

therealpxc commented:
I was imagining something simpler and more generic, which would be adjusting things so that instead of taking a single URI to a `repomd.xml.gz` file, we take a list of them or something. That way the same thing would be usable on other RPM-based distros and not rely in particular on the `scl` tool. I definitely think preferring the SCL repos is a good idea either way, since they're semi-official.

I'm not sure what the best way to do things is, either. Maybe we should just each do a low-effort implementation, put them up here on GitHub, and then compare notes to decide where to go from there?

FRidh commented:
@therealpxc @bhipple please discuss that at https://github.com/NixOS/nixpkgs/issues/32320

FRidh commented:
What do you think of including the init scripts and such in the Nix build? E.g., when wants to install the `binaryTarball` those scripts may be useful. Note this can be done after merging of this PR.

What is blocking this from being merged? Aside from the merge conflict there don't seem to be any more issues that cannot be looked at later.

gilligan commented:
This PR was started *over a year* ago ;( Meanwhile i need to install nix on some debian machine and i have to google for instructions on X manual steps to follow..

bhipple commented:
It would be really nice if we could get this rebased and merged before nix 2.0.  @abbradar it looks like all of the PR comments have been addressed, so maybe we can get it in?

While most retail users are going to start with either single-user nix or NixOS, most enterprise users are going to cautiously dip their toes into the nix ecosystem by installing a multi-user RPM or Deb.  If they can do so with one familiar command, it makes a much more favorable impression.

matthewbauer commented:
:+1: for merging before nix 2.0. I'm not sure if @abbradar is still around?

I have a branch here that is rebased onto HEAD that seems to be working:

https://github.com/matthewbauer/nix/tree/stable

abbradar commented:
Sorry for the long absence -- I had health problems and related issues for several months. I slowly get back to contributing and will have time to rebase and test this during this weekend!

FRidh commented:
I'm thinking it's maybe actually better to have these expressions elsewhere. Considering this repo is for the development of the package manager itself, I can imagine these files to just be noise, increasing the maintenance effort. Instead, if (some of) these expressions would be kept elsewhere, then that effort could focus on providing more features, as is done here.

abbradar commented:
As long as it's built and tested on Hydra along with actual Nix package it's not a bad idea I think to ease load on Nix maintainers. @edolstra, perhaps your opinion?

FRidh commented:
Exactly. I thought of having a folder in Nixpkgs for this.

abbradar commented:
Nixpkgs may be a bad idea because `nixUnstable` is bumped only sometimes and we'd want to catch breakages in those packages introduced by Nix as fast as possible, so a separate repo with a Hydra job and Nix repository as a second source may serve better.

gilligan commented:
We are talking about creating binary distributions for different linux platforms. I don’t understand why this should *not* be part of the nix project itself.

Why not even have GitHub releases and provide the various binary packages as downloads there.

The maintenance in general should be less of an issue once the currently ongoing reorganization talks materialize into something.

I think we should offer as much convenience and discoverability for installing nix as possible to provide new users a smooth start.

abbradar commented:
I've rebased the patches but to fully test I need to spin up several VMs manually and also update Ubuntu images in nixpkgs. Stay tuned!

Meanwhile @edolstra, do you have more comments on this? Are there any unhandled problems?

edolstra commented on src/libexpr/local.mk:
```diff
@@ -26,8 +26,6 @@ $(d)/parser-tab.cc $(d)/parser-tab.hh: $(d)/parser.y
 $(d)/lexer-tab.cc $(d)/lexer-tab.hh: $(d)/lexer.l
 	$(trace-gen) flex --outfile $(libexpr_DIR)/lexer-tab.cc --header-file=$(libexpr_DIR)/lexer-tab.hh $<
 
-clean-files += $(d)/parser-tab.cc $(d)/parser-tab.hh $(d)/lexer-tab.cc $(d)/lexer-tab.hh
```
Why this?

abbradar replied:
So that our source tarball would include generated lexer and parser -- otherwise it gives problems with various flex/bison versions (I remember some with CentOS back when we built for it, maybe there's more).

edolstra replied:
Our source tarball already includes the generated lexer and parser, see `dist-files` below.

abbradar replied:
I remember the actual reason now -- it's because `make clean` isn't supposed to remove anything from a freshly unpacked tarball. It breaks Debian build for instance.

abbradar commented:
I've updated distributions that our packages are built against. Unfortunately Ubuntu 16.04 build is currently broken on master so I've moved us towards 17.10. Good news: we now build against Fedora 27 and Debian 9 too.

To test this you'd need three nixpkgs patches:

https://github.com/NixOS/nixpkgs/pull/20767
https://github.com/NixOS/nixpkgs/pull/35176
https://github.com/NixOS/nixpkgs/pull/35177

Installation tests pass so we can be sure that at least binaries run and profile scripts work. To test nix-daemon units I'd need to spin up full-featured VMs; it'll take more time (perhaps one day we could do this with Nix too but it's too much work for now to build and controls complete VMs along with systemd -- surely there's not anything already usable that I missed? :3).

edolstra commented on scripts/nix-profile.sh.in:
```diff
@@ -60,6 +59,11 @@ if [ -n "$HOME" ] && [ -n "$USER" ]; then
 
+    # Try to route users through Nix daemon.
+    if [ "$USER" != "root" ] && [ -e "@localstatedir@/nix/daemon-socket/socket" ]; then
+        export NIX_REMOTE=daemon
```
Setting `NIX_REMOTE` should no longer be neceesary.

abbradar replied:
Fixed.

abbradar commented:
I've tested this in Ubuntu 17.10, Fedora 27 and Debian 9 VMs and it seem to work correctly completely with systemd unit. Pushed a Debian cleanup (`extraDebRequires` is no longer needed because runtime dependencies are specified in `debian/control` and partially derived by Debian package build system).

dtzWill commented:
Super excited about this!

Still going through but quick thoughts:

* Where possible, brotli (library, >= 1.0) as a build dep is nice so that "nix log" works on paths from official binary cache
* Ensure "key" tools are available on PATH so builtins don't fail to evaluate, see .. issue reference coming soon.

Mic92 commented:
Today I tested this on debian:

```console
$ nix-build release.nix -A deb_debian9x86_64 --arg nixpkgs ~/git/nixpkgs
```

However build users are not added at the moment by the package.

```
groupadd -r nixbld
for n in $(seq 1 10); do useradd -c "Nix build user $n" \
    -d /var/empty -g nixbld -G nixbld -M -N -r -s "$(which nologin)" \
    nixbld$n; done
```

abbradar commented:
I've rebased this against master again.

@dtzWill Sadly Brotli is available only on Fedora for now. As for needed binaries I'd prefer for #1345 be finished first so that `configure.ac` scripts discover and use those tools.

@Mic92 That's strange, I have it working. Do you have https://github.com/NixOS/nixpkgs/pull/21531 applied in your fork?

Mic92 commented:
@abbradar no, just this pull request.

abbradar commented:
@edolstra anything else with this? I so much hope I don't have to rebase it for another several months ~_~

dezgeg commented:
I think this approach of using rpmbuild and dpkg-buildpackage / debhelper sounds maintenance-heavy since we now have three build scripts to keep in sync. So I've created an alternative approach which uses the Nix-build version of Nix packaged with fpm: https://github.com/dezgeg/nix-fpm-multiuser

No pull request just yet but I guess the ML post works for any comments on that approach: https://groups.google.com/forum/#!topic/nix-devel/kIxMWOwX4m8

Mic92 commented:
@dezgeg I move the discussion out of this pull request here: https://github.com/dezgeg/nix-fpm-multiuser/issues/3

abbradar commented:
I like fpm-multiuser approach more and I've used it in production for some time. Let's close this.
